# *Data Preparation with Python* - Bagian 2
Pertanyaan, koreksi, maupun saran dapat disampaikan ke galuh.tunggadewi[at]gmail.com atau disampaikan dengan cara membuka [isu baru di GitHub](https://github.com/galuhsahid/data-preparation-with-python). Lisensi dapat dilihat [di sini](https://creativecommons.org/licenses/by-nc-sa/4.0/).

In [1]:
import pandas as pd

# Menghapus Data

In [147]:
df_aps = pd.read_csv("data/angka_partisipasi_sekolah.csv")

In [148]:
df_aps.head()

_id  idx                     vervar                              var  \
0    1    0  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
1    2    1  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
2    3    2  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
3    4    3  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
4    5    4  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   

          turvar  tahun  nilai  
0  Sekolah Dasar   2010  96.34  
1  Sekolah Dasar   2011  94.17  
2  Sekolah Dasar   2012  95.45  
3  Sekolah Dasar   2013    NaN  
4  Sekolah Dasar   2014  98.22

## Menghapus kolom

Ada dua metode:
- `del`
- `drop`

### del

In [149]:
del df_aps["_id"]

In [150]:
df_aps.head()

idx                     vervar                              var  \
0    0  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
1    1  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
2    2  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
3    3  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
4    4  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   

          turvar  tahun  nilai  
0  Sekolah Dasar   2010  96.34  
1  Sekolah Dasar   2011  94.17  
2  Sekolah Dasar   2012  95.45  
3  Sekolah Dasar   2013    NaN  
4  Sekolah Dasar   2014  98.22

⚠️ *Keyword* `del` akan menghapus kolom secara langsung (*in-place*). Apabila Anda ingin menggunakan `del`, pastikan bahwa Anda menghapus kolom yang benar. Apabila Anda ingin mencoba-coba dulu tanpa menghapus data yang sebenarnya, gunakan metode `drop` tanpa menempatkan *dataframe* Anda ke sebuah variabel.

### drop

📖  Dokumentasi `pandas.drop`: [tautan](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop.html)

- `axis`: bernilai 0 atau `index` apabila ingin menghapus label dari indeks; bernilai 1 apabila ingin menghapus kolom.
- `inplace`: `True` apabila operasi akan dilakukan ditempat; `False` apabila sebaliknya.

In [151]:
df_aps.drop("idx", axis=1).head()

vervar                              var         turvar  \
0  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)  Sekolah Dasar   
1  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)  Sekolah Dasar   
2  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)  Sekolah Dasar   
3  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)  Sekolah Dasar   
4  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)  Sekolah Dasar   

   tahun  nilai  
0   2010  96.34  
1   2011  94.17  
2   2012  95.45  
3   2013    NaN  
4   2014  98.22

In [152]:
df_aps.head()

idx                     vervar                              var  \
0    0  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
1    1  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
2    2  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
3    3  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
4    4  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   

          turvar  tahun  nilai  
0  Sekolah Dasar   2010  96.34  
1  Sekolah Dasar   2011  94.17  
2  Sekolah Dasar   2012  95.45  
3  Sekolah Dasar   2013    NaN  
4  Sekolah Dasar   2014  98.22

In [153]:
df_aps_deleted = df_aps.drop("idx", axis=1).head()

In [154]:
df_aps_deleted.head()

vervar                              var         turvar  \
0  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)  Sekolah Dasar   
1  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)  Sekolah Dasar   
2  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)  Sekolah Dasar   
3  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)  Sekolah Dasar   
4  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)  Sekolah Dasar   

   tahun  nilai  
0   2010  96.34  
1   2011  94.17  
2   2012  95.45  
3   2013    NaN  
4   2014  98.22

In [155]:
df_aps.drop("idx", axis=1, inplace=True)

In [156]:
df_aps.head()

vervar                              var         turvar  \
0  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)  Sekolah Dasar   
1  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)  Sekolah Dasar   
2  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)  Sekolah Dasar   
3  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)  Sekolah Dasar   
4  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)  Sekolah Dasar   

   tahun  nilai  
0   2010  96.34  
1   2011  94.17  
2   2012  95.45  
3   2013    NaN  
4   2014  98.22

In [157]:
df_aps = df_aps.drop("var", axis=1)

In [158]:
df_aps.head()

vervar         turvar  tahun  nilai
0  Kabupaten Polewali Mandar  Sekolah Dasar   2010  96.34
1  Kabupaten Polewali Mandar  Sekolah Dasar   2011  94.17
2  Kabupaten Polewali Mandar  Sekolah Dasar   2012  95.45
3  Kabupaten Polewali Mandar  Sekolah Dasar   2013    NaN
4  Kabupaten Polewali Mandar  Sekolah Dasar   2014  98.22

### Latihan
Hapus tiga kolom (\_id, idx, dan var) dalam satu kode sekaligus

In [159]:
df_aps = pd.read_csv("data/angka_partisipasi_sekolah.csv")

#### Hint

Anda bisa lihat di [dokumentasinya](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop.html) bahwa metode `drop` tidak hanya menerima satu label saja, tapi juga menerima *list*.

#### Solusi 1

In [160]:
dropped_columns = ["_id", "idx", "var"]
df_aps = df_aps.drop(dropped_columns, axis=1)

In [161]:
df_aps.head()

vervar         turvar  tahun  nilai
0  Kabupaten Polewali Mandar  Sekolah Dasar   2010  96.34
1  Kabupaten Polewali Mandar  Sekolah Dasar   2011  94.17
2  Kabupaten Polewali Mandar  Sekolah Dasar   2012  95.45
3  Kabupaten Polewali Mandar  Sekolah Dasar   2013    NaN
4  Kabupaten Polewali Mandar  Sekolah Dasar   2014  98.22

#### Solusi 2

In [162]:
df_aps = pd.read_csv("data/angka_partisipasi_sekolah.csv")

In [163]:
df_aps.drop(dropped_columns, axis=1, inplace=True)

In [164]:
df_aps.head()

vervar         turvar  tahun  nilai
0  Kabupaten Polewali Mandar  Sekolah Dasar   2010  96.34
1  Kabupaten Polewali Mandar  Sekolah Dasar   2011  94.17
2  Kabupaten Polewali Mandar  Sekolah Dasar   2012  95.45
3  Kabupaten Polewali Mandar  Sekolah Dasar   2013    NaN
4  Kabupaten Polewali Mandar  Sekolah Dasar   2014  98.22

## Menghapus baris

### Menghapus baris berdasarkan indeks

In [165]:
df_aps.drop([0, 1])

vervar                    turvar  tahun  nilai
2   Kabupaten Polewali Mandar             Sekolah Dasar   2012  95.45
3   Kabupaten Polewali Mandar             Sekolah Dasar   2013    NaN
4   Kabupaten Polewali Mandar             Sekolah Dasar   2014  98.22
5   Kabupaten Polewali Mandar             Sekolah Dasar   2015  97.04
6   Kabupaten Polewali Mandar  Sekolah Menengah Pertama   2010    NaN
7   Kabupaten Polewali Mandar  Sekolah Menengah Pertama   2011    NaN
8   Kabupaten Polewali Mandar  Sekolah Menengah Pertama   2012  80.02
9   Kabupaten Polewali Mandar  Sekolah Menengah Pertama   2013  80.04
10  Kabupaten Polewali Mandar  Sekolah Menengah Pertama   2014     --
11  Kabupaten Polewali Mandar  Sekolah Menengah Pertama   2015  90.51
12  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2010  32.79
13  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2011  55.88
14  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2012  56.01
15  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2013    NaN
16  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2014  61.73
17  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2014  64.32
18  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2015      -
19  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2015  66.26
20  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2015  66.26

### Menghapus baris berdasarkan nilai di sebuah kolom
Menghapus baris di mana `turvar` = 'Sekolah Dasar':

In [166]:
df_aps[df_aps.turvar != 'Sekolah Dasar']

vervar                    turvar  tahun  nilai
6   Kabupaten Polewali Mandar  Sekolah Menengah Pertama   2010    NaN
7   Kabupaten Polewali Mandar  Sekolah Menengah Pertama   2011    NaN
8   Kabupaten Polewali Mandar  Sekolah Menengah Pertama   2012  80.02
9   Kabupaten Polewali Mandar  Sekolah Menengah Pertama   2013  80.04
10  Kabupaten Polewali Mandar  Sekolah Menengah Pertama   2014     --
11  Kabupaten Polewali Mandar  Sekolah Menengah Pertama   2015  90.51
12  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2010  32.79
13  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2011  55.88
14  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2012  56.01
15  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2013    NaN
16  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2014  61.73
17  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2014  64.32
18  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2015      -
19  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2015  66.26
20  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2015  66.26

### Menghapus baris berdasarkan nomor baris

Pada contoh di bawah, kita akan menghapus baris pertama atau baris ke-0:

In [167]:
df_aps.drop(df_aps.index[0])

vervar                    turvar  tahun  nilai
1   Kabupaten Polewali Mandar             Sekolah Dasar   2011  94.17
2   Kabupaten Polewali Mandar             Sekolah Dasar   2012  95.45
3   Kabupaten Polewali Mandar             Sekolah Dasar   2013    NaN
4   Kabupaten Polewali Mandar             Sekolah Dasar   2014  98.22
5   Kabupaten Polewali Mandar             Sekolah Dasar   2015  97.04
6   Kabupaten Polewali Mandar  Sekolah Menengah Pertama   2010    NaN
7   Kabupaten Polewali Mandar  Sekolah Menengah Pertama   2011    NaN
8   Kabupaten Polewali Mandar  Sekolah Menengah Pertama   2012  80.02
9   Kabupaten Polewali Mandar  Sekolah Menengah Pertama   2013  80.04
10  Kabupaten Polewali Mandar  Sekolah Menengah Pertama   2014     --
11  Kabupaten Polewali Mandar  Sekolah Menengah Pertama   2015  90.51
12  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2010  32.79
13  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2011  55.88
14  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2012  56.01
15  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2013    NaN
16  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2014  61.73
17  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2014  64.32
18  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2015      -
19  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2015  66.26
20  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2015  66.26

Di contoh berikutnya, kita akan menghapus baris ketiga dan keempat.

⚠️  Karena pandas mulai menghitung dari 0, maka:
- Baris ke-3: 2
- Baris ke-4: 3

In [168]:
df_aps.head()

vervar         turvar  tahun  nilai
0  Kabupaten Polewali Mandar  Sekolah Dasar   2010  96.34
1  Kabupaten Polewali Mandar  Sekolah Dasar   2011  94.17
2  Kabupaten Polewali Mandar  Sekolah Dasar   2012  95.45
3  Kabupaten Polewali Mandar  Sekolah Dasar   2013    NaN
4  Kabupaten Polewali Mandar  Sekolah Dasar   2014  98.22

In [169]:
df_aps.drop(df_aps.index[[2, 3]])

vervar                    turvar  tahun  nilai
0   Kabupaten Polewali Mandar             Sekolah Dasar   2010  96.34
1   Kabupaten Polewali Mandar             Sekolah Dasar   2011  94.17
4   Kabupaten Polewali Mandar             Sekolah Dasar   2014  98.22
5   Kabupaten Polewali Mandar             Sekolah Dasar   2015  97.04
6   Kabupaten Polewali Mandar  Sekolah Menengah Pertama   2010    NaN
7   Kabupaten Polewali Mandar  Sekolah Menengah Pertama   2011    NaN
8   Kabupaten Polewali Mandar  Sekolah Menengah Pertama   2012  80.02
9   Kabupaten Polewali Mandar  Sekolah Menengah Pertama   2013  80.04
10  Kabupaten Polewali Mandar  Sekolah Menengah Pertama   2014     --
11  Kabupaten Polewali Mandar  Sekolah Menengah Pertama   2015  90.51
12  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2010  32.79
13  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2011  55.88
14  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2012  56.01
15  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2013    NaN
16  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2014  61.73
17  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2014  64.32
18  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2015      -
19  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2015  66.26
20  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2015  66.26

### Menghapus baris relatif terhadap urutannya dari baris terakhir

In [170]:
df_aps.tail()

vervar                 turvar  tahun  nilai
16  Kabupaten Polewali Mandar  Sekolah Menengah Umum   2014  61.73
17  Kabupaten Polewali Mandar  Sekolah Menengah Umum   2014  64.32
18  Kabupaten Polewali Mandar  Sekolah Menengah Umum   2015      -
19  Kabupaten Polewali Mandar  Sekolah Menengah Umum   2015  66.26
20  Kabupaten Polewali Mandar  Sekolah Menengah Umum   2015  66.26

In [171]:
df_aps.drop(df_aps.index[-2])

vervar                    turvar  tahun  nilai
0   Kabupaten Polewali Mandar             Sekolah Dasar   2010  96.34
1   Kabupaten Polewali Mandar             Sekolah Dasar   2011  94.17
2   Kabupaten Polewali Mandar             Sekolah Dasar   2012  95.45
3   Kabupaten Polewali Mandar             Sekolah Dasar   2013    NaN
4   Kabupaten Polewali Mandar             Sekolah Dasar   2014  98.22
5   Kabupaten Polewali Mandar             Sekolah Dasar   2015  97.04
6   Kabupaten Polewali Mandar  Sekolah Menengah Pertama   2010    NaN
7   Kabupaten Polewali Mandar  Sekolah Menengah Pertama   2011    NaN
8   Kabupaten Polewali Mandar  Sekolah Menengah Pertama   2012  80.02
9   Kabupaten Polewali Mandar  Sekolah Menengah Pertama   2013  80.04
10  Kabupaten Polewali Mandar  Sekolah Menengah Pertama   2014     --
11  Kabupaten Polewali Mandar  Sekolah Menengah Pertama   2015  90.51
12  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2010  32.79
13  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2011  55.88
14  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2012  56.01
15  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2013    NaN
16  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2014  61.73
17  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2014  64.32
18  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2015      -
20  Kabupaten Polewali Mandar     Sekolah Menengah Umum   2015  66.26

##  Mengubah nama kolom

📖  Dokumentasi `pandas.DataFrame.rename`: [tautan](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.rename.html)

In [172]:
df_aps = df_aps.rename(columns = {"vervar": "kabupaten",
                                  "turvar": "tingkat"})

In [173]:
df_aps.head()

kabupaten        tingkat  tahun  nilai
0  Kabupaten Polewali Mandar  Sekolah Dasar   2010  96.34
1  Kabupaten Polewali Mandar  Sekolah Dasar   2011  94.17
2  Kabupaten Polewali Mandar  Sekolah Dasar   2012  95.45
3  Kabupaten Polewali Mandar  Sekolah Dasar   2013    NaN
4  Kabupaten Polewali Mandar  Sekolah Dasar   2014  98.22

In [174]:
df_aps.columns

Index(['kabupaten', 'tingkat', 'tahun', 'nilai'], dtype='object')

# Menangani baris duplikat

### Mengidentifikasi baris duplikat dengan `pandas.DataFrame.duplicated`

Untuk menangani baris-baris yang merupakan duplikat, tentu kita perlu untuk mencari tahu apakah terdapat baris duplikat di data kita atau tidak.

📖  Dokumentasi `pandas.DataFrame.duplicated`: [tautan](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.duplicated.html)

In [175]:
df_aps = pd.read_csv("data/angka_partisipasi_sekolah.csv")

Metode `duplicated()` akan mengembalikan data bertipe *series* yang menunjukkan apakah baris tersebut benar merupakan sebuah duplikat atau bukan.

In [176]:
df_aps.duplicated()

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12    False
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20     True
dtype: bool

Pada contoh di atas, terlihat bahwa baris terakhir merupakan sebuah duplikat. Artinya, ada baris yang sama persis dengan baris tersebut muncul di baris-baris sebelumnya.

Untuk mencari tahu baris apakah yang merupakan sebuah duplikat, masukkan *series* tersebut ke operator `[]` untuk memilih baris-baris di mana `duplicated()` = `True`.

In [177]:
df_aps[df_aps.duplicated()]

_id  idx                     vervar                              var  \
20   19   17  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   

                   turvar  tahun  nilai  
20  Sekolah Menengah Umum   2015  66.26

Untuk melihat semua baris yang duplikat (termasuk baris yang muncul pertama kali), gunakan `keep`. `keep` menunjukkan baris-baris mana sajakah yang harus ditandai sebagai duplikat. `keep` dapat bernilai "first", "last", atau `False`.
- `keep` = "first": seluruh duplikat, kecuali baris yang muncul pertama, akan ditandai sebagai duplikat
- `keep` = "last": seluruh duplikat, kecuali baris yang muncul terakhir, akan ditandai sebagai duplikat
- `keep` = False: seluruh duplikat akan ditandai sebagai duplikat

In [178]:
df_aps[df_aps.duplicated(keep=False)]

_id  idx                     vervar                              var  \
19   19   17  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
20   19   17  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   

                   turvar  tahun  nilai  
19  Sekolah Menengah Umum   2015  66.26  
20  Sekolah Menengah Umum   2015  66.26

Seringkali, kita tidak membutuhkan semua kolom untuk menentukan apakah sebuah baris merupakan duplikat atau bukan. Sebagai contoh, pada dataset yang kita gunakan, mungkin tidak masuk akal apabila terdapat lebih dari satu data untuk kabupaten dan tingkat sekolah serta tahun yang sama, meksipun nilainya berbeda. Bisa jadi kita akan menganggap bahwa hanya salah satu nilai yang benar--misalnya, nilai di baris terakhir, dengan anggapan bahwa nilai di baris terakhir adalah nilai yang paling *up-to-date*.

In [115]:
df_aps[df_aps.duplicated(subset=["vervar", "turvar", "tahun"], keep=False)]

_id  idx                     vervar                              var  \
16   17   16  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
17   18   16  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
18   19   17  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
19   19   17  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
20   19   17  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   

                   turvar  tahun  nilai  
16  Sekolah Menengah Umum   2014  61.73  
17  Sekolah Menengah Umum   2014  64.32  
18  Sekolah Menengah Umum   2015      -  
19  Sekolah Menengah Umum   2015  66.26  
20  Sekolah Menengah Umum   2015  66.26

Pada contoh di atas, dengan bantuan `duplicated()` dan `keep=False`, kita dapat menemukan bahwa terdapat dua nilai yang berbeda untuk kabupaten, tingkat sekolah, dan tahun yang sama.

### Menghapus baris-baris duplikat dengan `pandas.DataFrame.drop_duplicates()`
pandas sudah menyediakan metode yang dapat langsung kita gunakan untuk menghapus baris-baris duplikat, yaitu `drop_duplicates()`.

📖  Dokumentasi `pandas.DataFrame.drop_duplicates`: [tautan](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.duplicated.html)

Secara *default*, `drop_duplicates()` menghapus baris-baris duplikat berdasarkan seluruh kolom, dengan parameter *default* `keep=First`. Jadi, seluruh baris duplikat akan dihapus kecuali baris yang muncul pertama kali.

In [128]:
df_aps_no_dup = df_aps.drop_duplicates()

Kita dapat menggunakan `duplicated()` untuk mengecek apakah masih ada baris-baris duplikat yang tersisa atau tidak.

In [129]:
df_aps_no_dup.duplicated()

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12    False
13    False
14    False
15    False
16    False
17    False
18    False
19    False
dtype: bool

In [130]:
df_aps_no_dup[df_aps_no_dup.duplicated()]

Empty DataFrame
Columns: [_id, idx, vervar, var, turvar, tahun, nilai]
Index: []

Meskipun kita sudah menghapus baris-baris duplikat, masih tersisa baris-baris yang duplikat berdasarkan definisi kolom vervar, turvar, dan tahun:

In [131]:
df_aps_no_dup[df_aps_no_dup.duplicated(subset=["vervar", "turvar", "tahun"])]

_id  idx                     vervar                              var  \
17   18   16  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
19   19   17  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   

                   turvar  tahun  nilai  
17  Sekolah Menengah Umum   2014  64.32  
19  Sekolah Menengah Umum   2015  66.26

In [132]:
df_aps_no_dup = df_aps.drop_duplicates(subset=["vervar", "turvar", "tahun"])

In [133]:
df_aps_no_dup[df_aps_no_dup.duplicated(subset=["vervar", "turvar", "tahun"])]

Empty DataFrame
Columns: [_id, idx, vervar, var, turvar, tahun, nilai]
Index: []

In [134]:
df_aps_no_dup

_id  idx                     vervar                              var  \
0     1    0  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
1     2    1  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
2     3    2  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
3     4    3  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
4     5    4  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
5     6    5  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
6     7    6  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
7     8    7  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
8     9    8  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
9    10    9  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
10   11   10  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
11   12   11  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
12   13   12  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
13   14   13  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
14   15   14  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
15   16   15  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
16   17   16  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
18   19   17  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   

                      turvar  tahun  nilai  
0              Sekolah Dasar   2010  96.34  
1              Sekolah Dasar   2011  94.17  
2              Sekolah Dasar   2012  95.45  
3              Sekolah Dasar   2013    NaN  
4              Sekolah Dasar   2014  98.22  
5              Sekolah Dasar   2015  97.04  
6   Sekolah Menengah Pertama   2010    NaN  
7   Sekolah Menengah Pertama   2011    NaN  
8   Sekolah Menengah Pertama   2012  80.02  
9   Sekolah Menengah Pertama   2013  80.04  
10  Sekolah Menengah Pertama   2014     --  
11  Sekolah Menengah Pertama   2015  90.51  
12     Sekolah Menengah Umum   2010  32.79  
13     Sekolah Menengah Umum   2011  55.88  
14     Sekolah Menengah Umum   2012  56.01  
15     Sekolah Menengah Umum   2013    NaN  
16     Sekolah Menengah Umum   2014  61.73  
18     Sekolah Menengah Umum   2015      -

`keep`
- `first` : Hilangkan baris-baris duplikat, kecuali baris pertama.
- `last` : Hilangkan baris-baris duplikat, kecuali baris terakhir.
- `False` : Hilangkan semua baris-baris duplikat.



In [135]:
df_aps.tail()

_id  idx                     vervar                              var  \
16   17   16  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
17   18   16  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
18   19   17  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
19   19   17  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
20   19   17  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   

                   turvar  tahun  nilai  
16  Sekolah Menengah Umum   2014  61.73  
17  Sekolah Menengah Umum   2014  64.32  
18  Sekolah Menengah Umum   2015      -  
19  Sekolah Menengah Umum   2015  66.26  
20  Sekolah Menengah Umum   2015  66.26

In [136]:
df_aps.drop_duplicates(subset=["vervar", "turvar", "tahun"], keep="last").tail()

_id  idx                     vervar                              var  \
13   14   13  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
14   15   14  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
15   16   15  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
17   18   16  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
20   19   17  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   

                   turvar  tahun  nilai  
13  Sekolah Menengah Umum   2011  55.88  
14  Sekolah Menengah Umum   2012  56.01  
15  Sekolah Menengah Umum   2013    NaN  
17  Sekolah Menengah Umum   2014  64.32  
20  Sekolah Menengah Umum   2015  66.26

In [55]:
df_aps.drop_duplicates(subset=["vervar", "turvar", "tahun"], keep=False)

_id  idx                     vervar                              var  \
0     1    0  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
1     2    1  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
2     3    2  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
3     4    3  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
4     5    4  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
5     6    5  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
6     7    6  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
7     8    7  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
8     9    8  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
9    10    9  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
10   11   10  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
11   12   11  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
12   13   12  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
13   14   13  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
14   15   14  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
15   16   15  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   

                      turvar  tahun  nilai  
0              Sekolah Dasar   2010  96.34  
1              Sekolah Dasar   2011  94.17  
2              Sekolah Dasar   2012  95.45  
3              Sekolah Dasar   2013    NaN  
4              Sekolah Dasar   2014  98.22  
5              Sekolah Dasar   2015  97.04  
6   Sekolah Menengah Pertama   2010    NaN  
7   Sekolah Menengah Pertama   2011    NaN  
8   Sekolah Menengah Pertama   2012  80.02  
9   Sekolah Menengah Pertama   2013  80.04  
10  Sekolah Menengah Pertama   2014     --  
11  Sekolah Menengah Pertama   2015  90.51  
12     Sekolah Menengah Umum   2010  32.79  
13     Sekolah Menengah Umum   2011  55.88  
14     Sekolah Menengah Umum   2012  56.01  
15     Sekolah Menengah Umum   2013    NaN

# Meng-*handle missing values*
*Missing values* bisa terjadi karena:
- Pengguna lupa untuk mengisi *field*.
- Data hilang ketika melakukan integrasi/transfer.
- Ada *error* dalam kode.

Penting untuk mengetahui alasan mengapa sebuah data dapat hilang, karena ini dapat memengaruhi metode pengisian *missing data*. 

Lebih lanjut tentang ini: [tautan](https://github.com/matthewbrems/ODSC-missing-data-may-18/blob/master/Analysis%20with%20Missing%20Data.pdf?source=post_page---------------------------)

In [179]:
df_aps = pd.read_csv("data/angka_partisipasi_sekolah.csv")

### *Missing value* dengan bentuk standar

Pandas menganggap cell yang kosong maupun cell berisi "NA" sebagai cell yang kosong. Anda bisa melihat di *file* `data/angka_partisipasi_sekolah.csv` bahwa data nilai di Kabupaten Polewali Mandar untuk tingkat Sekolah Menengah Umum di tahun 2013 tertulis "NA". Apabila Anda lihat pada tabel di bawah, pandas dapat langsung membacanya sebagai sebuah *missing value* (NaN).

In [180]:
df_aps

_id  idx                     vervar                              var  \
0     1    0  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
1     2    1  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
2     3    2  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
3     4    3  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
4     5    4  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
5     6    5  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
6     7    6  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
7     8    7  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
8     9    8  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
9    10    9  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
10   11   10  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
11   12   11  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
12   13   12  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
13   14   13  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
14   15   14  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
15   16   15  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
16   17   16  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
17   18   16  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
18   19   17  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
19   19   17  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
20   19   17  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   

                      turvar  tahun  nilai  
0              Sekolah Dasar   2010  96.34  
1              Sekolah Dasar   2011  94.17  
2              Sekolah Dasar   2012  95.45  
3              Sekolah Dasar   2013    NaN  
4              Sekolah Dasar   2014  98.22  
5              Sekolah Dasar   2015  97.04  
6   Sekolah Menengah Pertama   2010    NaN  
7   Sekolah Menengah Pertama   2011    NaN  
8   Sekolah Menengah Pertama   2012  80.02  
9   Sekolah Menengah Pertama   2013  80.04  
10  Sekolah Menengah Pertama   2014     --  
11  Sekolah Menengah Pertama   2015  90.51  
12     Sekolah Menengah Umum   2010  32.79  
13     Sekolah Menengah Umum   2011  55.88  
14     Sekolah Menengah Umum   2012  56.01  
15     Sekolah Menengah Umum   2013    NaN  
16     Sekolah Menengah Umum   2014  61.73  
17     Sekolah Menengah Umum   2014  64.32  
18     Sekolah Menengah Umum   2015      -  
19     Sekolah Menengah Umum   2015  66.26  
20     Sekolah Menengah Umum   2015  66.26

In [181]:
df_aps[df_aps["nilai"].isnull()]

_id  idx                     vervar                              var  \
3     4    3  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
6     7    6  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
7     8    7  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
15   16   15  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   

                      turvar  tahun nilai  
3              Sekolah Dasar   2013   NaN  
6   Sekolah Menengah Pertama   2010   NaN  
7   Sekolah Menengah Pertama   2011   NaN  
15     Sekolah Menengah Umum   2013   NaN

Terkadang, pada data yang kita miliki, *missing value* tidak ditandai dengan "NA" atau tidak dibiarkan kosong, melainkan ditandai dengan tanda "-". Kita dapat memberitahu pandas untuk membaca *cell* berisi karakter "-" sebagai *missing value* dengan mendefinisikan `na_values` pada `read_csv()`.

Pada contoh di bawah, kita memberitahu pandas untuk membaca *cell* berisi karakter "-" dan "--" sebagai *missing value*.

In [182]:
missing_values = ["-", "--"]
df_aps = pd.read_csv("data/angka_partisipasi_sekolah.csv", na_values=missing_values)

In [183]:
df_aps[df_aps["nilai"].isnull()]

_id  idx                     vervar                              var  \
3     4    3  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
6     7    6  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
7     8    7  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
10   11   10  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
15   16   15  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
18   19   17  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   

                      turvar  tahun  nilai  
3              Sekolah Dasar   2013    NaN  
6   Sekolah Menengah Pertama   2010    NaN  
7   Sekolah Menengah Pertama   2011    NaN  
10  Sekolah Menengah Pertama   2014    NaN  
15     Sekolah Menengah Umum   2013    NaN  
18     Sekolah Menengah Umum   2015    NaN

Menghitung berapa jumlah *missing value* di masing-masing kolom:

In [184]:
df_aps.isnull().sum()

_id       0
idx       0
vervar    0
var       0
turvar    0
tahun     0
nilai     6
dtype: int64

Mencari tahu apakah ada *missing value* di antara semua kolom yang ada pada data:

In [185]:
df_aps.isnull().values.any()

True

Mencari tahu berapakah *missing value* yang terdapat di antara semua kolom yang ada pada data:

In [186]:
df_aps.isnull().sum().sum()

6

Selanjutnya, kita akan membahas berbagai metode untuk pengisian *missing value* dan bagaimana kita dapat mengimplementasikannya menggunakan pandas.

⚠️ Metode pengisian *missing value* yang sebaiknya digunakan bergantung pada data Anda serta konteks analisis yang ingin Anda lakukan. Pembahasan tentang hal tersebut berada di luar *scope* dari *notebook* ini, namun Anda dapat merujuk ke [tautan ini](https://github.com/matthewbrems/ODSC-missing-data-may-18/blob/master/Analysis%20with%20Missing%20Data.pdf?source=post_page---------------------------) untuk mencari tahu lebih lanjut.

### Mengganti *missing value* dengan sebuah nilai

In [207]:
missing_values = ["-", "--"]
df_aps = pd.read_csv("data/angka_partisipasi_sekolah.csv", na_values=missing_values)

In [187]:
df_aps.head()

_id  idx                     vervar                              var  \
0    1    0  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
1    2    1  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
2    3    2  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
3    4    3  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
4    5    4  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   

          turvar  tahun  nilai  
0  Sekolah Dasar   2010  96.34  
1  Sekolah Dasar   2011  94.17  
2  Sekolah Dasar   2012  95.45  
3  Sekolah Dasar   2013    NaN  
4  Sekolah Dasar   2014  98.22

Mengganti semua *missing value* dengan nilai 125:

In [188]:
df_aps.fillna(125)

_id  idx                     vervar                              var  \
0     1    0  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
1     2    1  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
2     3    2  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
3     4    3  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
4     5    4  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
5     6    5  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
6     7    6  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
7     8    7  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
8     9    8  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
9    10    9  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
10   11   10  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
11   12   11  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
12   13   12  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
13   14   13  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
14   15   14  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
15   16   15  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
16   17   16  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
17   18   16  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
18   19   17  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
19   19   17  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
20   19   17  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   

                      turvar  tahun   nilai  
0              Sekolah Dasar   2010   96.34  
1              Sekolah Dasar   2011   94.17  
2              Sekolah Dasar   2012   95.45  
3              Sekolah Dasar   2013  125.00  
4              Sekolah Dasar   2014   98.22  
5              Sekolah Dasar   2015   97.04  
6   Sekolah Menengah Pertama   2010  125.00  
7   Sekolah Menengah Pertama   2011  125.00  
8   Sekolah Menengah Pertama   2012   80.02  
9   Sekolah Menengah Pertama   2013   80.04  
10  Sekolah Menengah Pertama   2014  125.00  
11  Sekolah Menengah Pertama   2015   90.51  
12     Sekolah Menengah Umum   2010   32.79  
13     Sekolah Menengah Umum   2011   55.88  
14     Sekolah Menengah Umum   2012   56.01  
15     Sekolah Menengah Umum   2013  125.00  
16     Sekolah Menengah Umum   2014   61.73  
17     Sekolah Menengah Umum   2014   64.32  
18     Sekolah Menengah Umum   2015  125.00  
19     Sekolah Menengah Umum   2015   66.26  
20     Sekolah Menengah Umum   2015   66.26

### Mengganti *missing value* dengan sebuah nilai pada lokasi tertentu

In [198]:
missing_values = ["-", "--"]
df_aps = pd.read_csv("data/angka_partisipasi_sekolah.csv", na_values=missing_values)

Mengganti *missing value* dengan indeks `3` menjadi 129:

In [199]:
df_aps.loc[3, "nilai"] = 129

In [200]:
df_aps

_id  idx                     vervar                              var  \
0     1    0  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
1     2    1  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
2     3    2  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
3     4    3  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
4     5    4  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
5     6    5  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
6     7    6  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
7     8    7  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
8     9    8  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
9    10    9  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
10   11   10  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
11   12   11  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
12   13   12  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
13   14   13  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
14   15   14  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
15   16   15  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
16   17   16  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
17   18   16  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
18   19   17  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
19   19   17  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
20   19   17  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   

                      turvar  tahun   nilai  
0              Sekolah Dasar   2010   96.34  
1              Sekolah Dasar   2011   94.17  
2              Sekolah Dasar   2012   95.45  
3              Sekolah Dasar   2013  129.00  
4              Sekolah Dasar   2014   98.22  
5              Sekolah Dasar   2015   97.04  
6   Sekolah Menengah Pertama   2010     NaN  
7   Sekolah Menengah Pertama   2011     NaN  
8   Sekolah Menengah Pertama   2012   80.02  
9   Sekolah Menengah Pertama   2013   80.04  
10  Sekolah Menengah Pertama   2014     NaN  
11  Sekolah Menengah Pertama   2015   90.51  
12     Sekolah Menengah Umum   2010   32.79  
13     Sekolah Menengah Umum   2011   55.88  
14     Sekolah Menengah Umum   2012   56.01  
15     Sekolah Menengah Umum   2013     NaN  
16     Sekolah Menengah Umum   2014   61.73  
17     Sekolah Menengah Umum   2014   64.32  
18     Sekolah Menengah Umum   2015     NaN  
19     Sekolah Menengah Umum   2015   66.26  
20     Sekolah Menengah Umum   2015   66.26

### Mengganti *missing value* dengan *mean* dari kolom

In [208]:
missing_values = ["-", "--"]
df_aps = pd.read_csv("data/angka_partisipasi_sekolah.csv", na_values=missing_values)

In [209]:
mean = df_aps['nilai'].mean()

In [210]:
mean

75.66933333333333

In [211]:
df_aps['nilai'] = df_aps['nilai'].fillna(mean)

In [212]:
df_aps

_id  idx                     vervar                              var  \
0     1    0  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
1     2    1  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
2     3    2  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
3     4    3  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
4     5    4  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
5     6    5  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
6     7    6  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
7     8    7  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
8     9    8  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
9    10    9  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
10   11   10  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
11   12   11  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
12   13   12  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
13   14   13  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
14   15   14  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
15   16   15  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
16   17   16  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
17   18   16  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
18   19   17  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
19   19   17  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
20   19   17  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   

                      turvar  tahun      nilai  
0              Sekolah Dasar   2010  96.340000  
1              Sekolah Dasar   2011  94.170000  
2              Sekolah Dasar   2012  95.450000  
3              Sekolah Dasar   2013  75.669333  
4              Sekolah Dasar   2014  98.220000  
5              Sekolah Dasar   2015  97.040000  
6   Sekolah Menengah Pertama   2010  75.669333  
7   Sekolah Menengah Pertama   2011  75.669333  
8   Sekolah Menengah Pertama   2012  80.020000  
9   Sekolah Menengah Pertama   2013  80.040000  
10  Sekolah Menengah Pertama   2014  75.669333  
11  Sekolah Menengah Pertama   2015  90.510000  
12     Sekolah Menengah Umum   2010  32.790000  
13     Sekolah Menengah Umum   2011  55.880000  
14     Sekolah Menengah Umum   2012  56.010000  
15     Sekolah Menengah Umum   2013  75.669333  
16     Sekolah Menengah Umum   2014  61.730000  
17     Sekolah Menengah Umum   2014  64.320000  
18     Sekolah Menengah Umum   2015  75.669333  
19     Sekolah Menengah Umum   2015  66.260000  
20     Sekolah Menengah Umum   2015  66.260000

### Mengganti *missing value* dengan nilai di baris berikutnya

In [213]:
missing_values = ["-", "--"]
df_aps = pd.read_csv("data/angka_partisipasi_sekolah.csv", na_values=missing_values)

Kita dapat mengganti *missing value* dengan nilai di baris berikutnya. Sebagai contoh, pada tabel di bawah ini, kolom nilai di baris ke-4 (indeks `3`) akan diisi dengan nilai yang ada di baris berikutnya, yaitu 98.22.

In [216]:
df_aps

_id  idx                     vervar                              var  \
0     1    0  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
1     2    1  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
2     3    2  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
3     4    3  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
4     5    4  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
5     6    5  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
6     7    6  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
7     8    7  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
8     9    8  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
9    10    9  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
10   11   10  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
11   12   11  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
12   13   12  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
13   14   13  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
14   15   14  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
15   16   15  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
16   17   16  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
17   18   16  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
18   19   17  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
19   19   17  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
20   19   17  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   

                      turvar  tahun  nilai  
0              Sekolah Dasar   2010  96.34  
1              Sekolah Dasar   2011  94.17  
2              Sekolah Dasar   2012  95.45  
3              Sekolah Dasar   2013    NaN  
4              Sekolah Dasar   2014  98.22  
5              Sekolah Dasar   2015  97.04  
6   Sekolah Menengah Pertama   2010    NaN  
7   Sekolah Menengah Pertama   2011    NaN  
8   Sekolah Menengah Pertama   2012  80.02  
9   Sekolah Menengah Pertama   2013  80.04  
10  Sekolah Menengah Pertama   2014    NaN  
11  Sekolah Menengah Pertama   2015  90.51  
12     Sekolah Menengah Umum   2010  32.79  
13     Sekolah Menengah Umum   2011  55.88  
14     Sekolah Menengah Umum   2012  56.01  
15     Sekolah Menengah Umum   2013    NaN  
16     Sekolah Menengah Umum   2014  61.73  
17     Sekolah Menengah Umum   2014  64.32  
18     Sekolah Menengah Umum   2015    NaN  
19     Sekolah Menengah Umum   2015  66.26  
20     Sekolah Menengah Umum   2015  66.26

In [217]:
df_aps.fillna(method ='bfill') 

_id  idx                     vervar                              var  \
0     1    0  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
1     2    1  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
2     3    2  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
3     4    3  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
4     5    4  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
5     6    5  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
6     7    6  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
7     8    7  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
8     9    8  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
9    10    9  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
10   11   10  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
11   12   11  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
12   13   12  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
13   14   13  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
14   15   14  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
15   16   15  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
16   17   16  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
17   18   16  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
18   19   17  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
19   19   17  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
20   19   17  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   

                      turvar  tahun  nilai  
0              Sekolah Dasar   2010  96.34  
1              Sekolah Dasar   2011  94.17  
2              Sekolah Dasar   2012  95.45  
3              Sekolah Dasar   2013  98.22  
4              Sekolah Dasar   2014  98.22  
5              Sekolah Dasar   2015  97.04  
6   Sekolah Menengah Pertama   2010  80.02  
7   Sekolah Menengah Pertama   2011  80.02  
8   Sekolah Menengah Pertama   2012  80.02  
9   Sekolah Menengah Pertama   2013  80.04  
10  Sekolah Menengah Pertama   2014  90.51  
11  Sekolah Menengah Pertama   2015  90.51  
12     Sekolah Menengah Umum   2010  32.79  
13     Sekolah Menengah Umum   2011  55.88  
14     Sekolah Menengah Umum   2012  56.01  
15     Sekolah Menengah Umum   2013  61.73  
16     Sekolah Menengah Umum   2014  61.73  
17     Sekolah Menengah Umum   2014  64.32  
18     Sekolah Menengah Umum   2015  66.26  
19     Sekolah Menengah Umum   2015  66.26  
20     Sekolah Menengah Umum   2015  66.26

Di awal, terdapat *missing value* di baris ke-7 (indeks `6`) dan baris ke-8 (indeks `7`). Ketika kita menggunakan `method ='bfill'`, baris ke-8 akan diisi terlebih dahulu mengikuti nilai di baris ke-9 (indeks `8`), dan kemudian baris ke-7 akan mengikuti nilai baru di baris ke-8.

### Latihan
Bagaimana caranya mengisi *missing value* di kolom nilai dengan median dari kolom tersebut?

In [219]:
missing_values = ["-", "--"]
df_aps = pd.read_csv("data/angka_partisipasi_sekolah.csv", na_values=missing_values)

#### Solusi

In [220]:
median = df_aps['nilai'].median()

In [221]:
median

80.02

In [222]:
df_aps['nilai'].fillna(median, inplace=True)

In [223]:
df_aps

_id  idx                     vervar                              var  \
0     1    0  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
1     2    1  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
2     3    2  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
3     4    3  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
4     5    4  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
5     6    5  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
6     7    6  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
7     8    7  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
8     9    8  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
9    10    9  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
10   11   10  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
11   12   11  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
12   13   12  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
13   14   13  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
14   15   14  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
15   16   15  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
16   17   16  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
17   18   16  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
18   19   17  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
19   19   17  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   
20   19   17  Kabupaten Polewali Mandar  Angka Partisipasi Sekolah (APS)   

                      turvar  tahun  nilai  
0              Sekolah Dasar   2010  96.34  
1              Sekolah Dasar   2011  94.17  
2              Sekolah Dasar   2012  95.45  
3              Sekolah Dasar   2013  80.02  
4              Sekolah Dasar   2014  98.22  
5              Sekolah Dasar   2015  97.04  
6   Sekolah Menengah Pertama   2010  80.02  
7   Sekolah Menengah Pertama   2011  80.02  
8   Sekolah Menengah Pertama   2012  80.02  
9   Sekolah Menengah Pertama   2013  80.04  
10  Sekolah Menengah Pertama   2014  80.02  
11  Sekolah Menengah Pertama   2015  90.51  
12     Sekolah Menengah Umum   2010  32.79  
13     Sekolah Menengah Umum   2011  55.88  
14     Sekolah Menengah Umum   2012  56.01  
15     Sekolah Menengah Umum   2013  80.02  
16     Sekolah Menengah Umum   2014  61.73  
17     Sekolah Menengah Umum   2014  64.32  
18     Sekolah Menengah Umum   2015  80.02  
19     Sekolah Menengah Umum   2015  66.26  
20     Sekolah Menengah Umum   2015  66.26

# *Checkpoint*
- Menghapus kolom (del, drop)
- Menghapus baris (berdasarkan indeks, nilai di kolom, nomor baris...)
- Mengubah nama kolom
- Menangani baris duplikat (identifikasi dan penghapusan)
- Meng-handle *missing values* (membaca berbagai jenis *missing value*, mengubah *missing value* menjadi sebuah nilai, *mean*, median, dll.)

![alt text](https://media3.giphy.com/media/12XDYvMJNcmLgQ/giphy.gif?cid=790b76115d4482dd58414a5049302fa6&rid=giphy.gif)